astropy/cosmology/funcs.py

```diff
@@ -255,3 +276,105 @@ def f(z):
+array([3.19812061, 0.75620443])
+
+``fval`` can be any shape:
+
+>>> z_at_value(Planck13.age, [[2, 7], [1, 3]] * u.Gyr)    # doctest: +FLOAT_CMP
+array([[3.19812061, 0.75620443],
+       [5.67661227, 2.19131955]])
+
+Other arguments can be vectorized. For non-monotic functions  -- for example,
+the angular diameter distance -- this can be useful to find all solutions.
+
+>>> z_at_value(Planck13.angular_diameter_distance, 1500 * u.Mpc,
+...            zmin=[0, 2.5], zmax=[2, 4])                # doctest: +FLOAT_CMP
+array([0.68127747, 3.79149062])
+
+The ``bracket`` argument can likewise be be a vector. However, since bracket must
+already be a sequence (or None), it MUST be given as an object `numpy.ndarray`.
+Importantly, the depth of the array must be such that each bracket subsequence
+is an object. Errors or unexpected results will happen otherwise.
+A convenient means to ensure the right depth is by including a length-0 tuple
+as a bracket and then truncating the object array to remove the placeholder.
+This can be seen in the following example:
+
+>>> bracket=np.array([(1.0, 1.2),(2.0, 2.5), ()], dtype=object)[:-1]
+>>> z_at_value(Planck18.angular_diameter_distance, 1500 * u.Mpc,
+...            bracket=bracket)  # doctest: +SKIP
+array([0.68044452, 3.7823268])
+"""
+
+
+@format_doc(_docstring_z_at_value + _docstring_z_at_value_vector_examples,
+            func_name="z_at_value", name_space="",
+            fval_type="scalar or vector",
+            z_type="float or array-like['dimensionless']",
+            maxfun_type="int or array-like",
+            bracket_type="sequence or object array[sequence]",
+            return_type="scalar or array", raises_docs=_raises_docs)
+def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500,
+               method='Brent', bracket=None, verbose=False):
+    """``z_at_scalar_value`` for vectorized inputs."""
+    # `fval` can be a Quantity, which isn't (yet) compatible w/ `numpy.nditer`
+    # so we strip it of units for broadcasting and restore the units when
+    # passing the elements to `z_at_scalar_value`.
+    fval = np.asanyarray(fval)
+    unit = getattr(fval, 'unit', 1)  # can be unitless
+
+    # bracket must be an object array (assumed to be correct) or a 'scalar'
+    # bracket: 2 or 3 elt sequence
+    if not isinstance(bracket, np.ndarray):  # 'scalar' bracket
+        if bracket is not None and len(bracket) not in (2, 3):
+            raise ValueError("`bracket` is not an array "
+                             "nor a 2 (or 3) element sequence.")
+        else:  # munge bracket into a 1-elt object array
+            bracket = np.array([bracket, ()], dtype=object)[:1].squeeze()
+    if bracket.dtype != np.object_:
+        raise TypeError(f"`bracket` has dtype {bracket.dtype}, not 'O'")
+
+    # make multi-dimensional iterator for all but `method`, `verbose`
+    # TODO: figure out 'reduce_ok', so scalar returns scalar
```

Comment:
I thought this was now OK!?

Reply:
It's fixed in that I manually check for a scalar and correct the output. I think there's some way to get 'reduce_ok' to do this automatically, but I can't get the examples in the numpy documentation to work here.

Reply:
Hence. I'm leaving the note for myself or an enterprising contributor. Low priority, but it would be nice.

Reply:
Ah, I thought it was the problem that the code returned a 1-D array rather than an array scalar. I think `reduce_ok` is really meant for dimensions that are reduced over (i.e., it tells the iterator that it is OK for a dimension to be missing or be unity in the output).

Reply:
In the first example in https://numpy.org/doc/stable/reference/arrays.nditer.html#reduction-iteration  it seems to work for degrading to a scalar.

Reply:
Indeed, and I did not know that. But I think it is not directly relevant, since you are not actually reducing over any dimension..